### DISCUSSION TITLE
Forward references in methods

### DISCUSSION THREAD

**Discussion by superbobry** [python/typing#1180] [Q&A] [answered]
Hi everyone,

How should a type checker resolve the return type of `f` in 

```
import abc

class A:
    def abc(self) -> 'abc.ABC': ...
```

?

According to [PEP 484](https://peps.python.org/pep-0484/#forward-references)

> The local and global namespace in which it [a forward reference] is evaluated should be the same namespaces in which default arguments to the same function would be evaluated.

which, I think, implies that the reference `abc` in the return type of `f` should be resolved to `abc` the method (in the class namespace) and *not* `abc` the module in the global namespace. Does that sound right? 

---

A few more motivating examples

```
class A:
    C = int
    def f(self) -> C: ...
```

```
class A:
    def f(self) -> C: ...
    C = int
```

**Comment by erictraut** [comment, marked as answer]
That's correct. Mypy and pyright both interpret it this way. However, I'd recommend against using conflicting names like this. It harms readability, and you will inevitably see inconsistencies across tools.

**Reply by superbobry** [reply]
Thanks Eric. I agree that it's best not to use names like this. 

Just to double check, are you saying that the correct behavior is emit a type error for the return type, because `abc` the method does not have an attribute `ABC`?

---

For completeness, the output of mypy, pyright and pytype executed on the original example with `reveal_type(A().abc())` added.  

mypy 0.950

```
/tmp/foo.py:6: note: Revealed type is "abc.ABC"
``` 

pyright 1.1.245

```
Found 1 source file
/tmp/foo.py
  /tmp/foo.py:6:13 - information: Type of "A().abc()" is "ABC"
```

pytype 2022.5.5

```
File "/tmp/foo.py", line 4, in abc: bad return type [bad-return-type]
           Expected: abc.ABC
  Actually returned: None
File "/tmp/foo.py", line 6, in <module>: abc.ABC [reveal-type]
```

**Reply by erictraut** [reply]
Apologies for the unclear answer. I think type checkers should evaluate the forward-declared type in this case as though `abc.ABC` refers to the `ABC` symbol in the `abc` module. It should not generate an error in this case. Mypy, pyright and pytype all appear to be in agreement here. If you evaluate the forward-declared type at runtime, you'll find that it's likewise in agreement.

**Reply by superbobry** [reply]
PEP 484 suggests that `abc` should be evaluated (or resolved) in the same namespace where the method is defined. I imagine a type checker could implement that by finding all assignments to `abc` reaching the exit from the class body. Clearly, this is not what existing type checkers do. Could you comment on how pyright does the evaluation in this case?

**Reply by erictraut** [reply]
Pyright used to do as you suggested, but it was [pointed out that this doesn't match the runtime behavior when evaluating forward references](https://github.com/microsoft/pyright/issues/2644). This created problems when working with runtime type checkers. I think all of the type checkers and the runtime are now in alignment. The runtime behavior is clearly documented in [PEP 563](https://peps.python.org/pep-0563/#implementation).

As I mentioned above, I encourage developers to avoid this type of namespace overlap because it makes the code fragile (e.g. when removing or adding a `from __future__ import annotations` to the file, the behavior changes in an unexpected way), and you may see different behaviors from different tools.

**Reply by superbobry** [reply]
Thanks Eric, this is very helpful.

One remaining issue is the discrepancy between PEP 563 and `typing.get_type_hints`. PEP 563 says that 

> For classes, localns can be composed by chaining vars of the given class and its base classes (in the method resolution order).

and then gives multiple examples of valid and invalid forward references, including

```
class A:
    T = int
    def method(self) -> 'T': ...
```

(I have adapted the second example from the ["Backwards compatibility"](https://peps.python.org/pep-0563/#backwards-compatibility) section)

which, however, does not work with `typing.get_type_hints`

```
>>> typing.get_type_hints(A.f)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.9/typing.py", line 1496, in get_type_hints
    value = _eval_type(value, globalns, localns)
  File "/usr/lib/python3.9/typing.py", line 292, in _eval_type
    return t._evaluate(globalns, localns, recursive_guard)
  File "/usr/lib/python3.9/typing.py", line 554, in _evaluate
    eval(self.__forward_code__, globalns, localns),
  File "<string>", line 1, in <module>
NameError: name 'T' is not defined
```

Could you advise how to think about that? What is the general rule for resolving a forward reference?

**Reply by hmc-cs-mdrissi** [reply]
The rule is you need to pass locals/globals yourself. typing.get_type_hints full signature is (obj, globalns=None, localns=None, include_extras=False).

For code that inspects type annotations inside of class/function scope it needs to keep track of right namespaces to pass in. I have some code that uses decorator for inspecting type annotations of a class and when that decorator is called it looks at caller's stack frame for it's namespaces and saves them for type evaluation later. Alternatively you could have a function that registers forward references in some dict that you reuse as namespace later. I use that to handle case for type aliases I need to evaluate in a different module later.

A lot of runtime type inspection libraries have edge cases for situations like this as managing it correctly is a bit messy and is easiest done by forcing explicit registration elsewhere.

**Reply by superbobry** [reply]
Thanks @hmc-cs-mdrissi and apologies for the confusion! My question was about the rule a type checker should use to implement evaluation of forward references.
